### ISSUE TITLE
`npm install -g flow-typed` fails on node v4.4.0 and npm 2.x

### ISSUE THREAD

**Issue by valscion** [flowtype/flow-typed#140]
Created a new issue from the comment on #4.

Running on OS X El Capitan, npm version 2.15.2 and node v4.4.0, I get this:

```
$ npm install flow-typed -g
-
> nodegit@0.13.2 install /Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit
> node lifecycleScripts/install

[nodegit] Fetching binary from S3.
[nodegit] Failed to install prebuilt binary:
{ [Error: Command failed: /bin/sh -c "/Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit/node_modules/.bin/node-pre-gyp" install --fallback-to-build=false
module.js:327
    throw err;
    ^

Error: Cannot find module 'nopt'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit/node_modules/node-pre-gyp/lib/node-pre-gyp.js:14:12)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
]
  killed: false,
  code: 1,
  signal: null,
  cmd: '/bin/sh -c "/Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit/node_modules/.bin/node-pre-gyp" install --fallback-to-build=false' }
[nodegit] Building manually. (You'll be here a while.)
[nodegit] Regenerating and configuring code
nodegit@0.13.2 /Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/nodegit
├── promisify-node@0.3.0 (nodegit-promise@4.0.0)
└── fs-extra@0.26.7 (path-is-absolute@1.0.0, klaw@1.3.0, graceful-fs@4.1.4, jsonfile@2.3.1, rimraf@2.5.2)

[nodegit] Configuring libssh2.
ERROR - Could not generate native code
[TypeError: _(...).forEach(...).value is not a function]
[nodegit] Everything is ready to go, attempting compilation
events.js:141
      throw er; // Unhandled 'error' event
      ^

Error: spawn /Users/vesa/.nvm/versions/node/v4.4.0/lib/node_modules/flow-typed/node_modules/nodegit/node_modules/.bin/node-gyp ENOENT
    at exports._errnoException (util.js:870:11)
    at Process.ChildProcess._handle.onexit (internal/child_process.js:178:32)
    at onErrorNT (internal/child_process.js:344:16)
    at nextTickCallbackWith2Args (node.js:442:9)
    at process._tickCallback (node.js:356:17)
npm ERR! Darwin 15.5.0
npm ERR! argv "/Users/vesa/.nvm/versions/node/v4.4.0/bin/node" "/Users/vesa/.nvm/versions/node/v4.4.0/bin/npm" "install" "flow-typed" "-g"
npm ERR! node v4.4.0
npm ERR! npm  v2.15.2
npm ERR! code ELIFECYCLE
```

**Comment by jeffmo** [comment]
Consolidating @ryyppy's comment from #4:

> @valscion Hmm... strange, so I could install it successfully with node version 4.0.0... then I tried it with 4.4.0 and it failed for me as well...

**Comment by ryyppy** [comment]
Looks like nodegit is the problem... and they seem to have a fix with version 0.13.1:
https://github.com/nodegit/nodegit/releases/tag/v0.13.1

**Comment by jeffmo** [comment]
@valscion Can you confirm that `rm -rf node_modules && npm install` (which should hopefully install the patch release) fixes the issue?

**Comment by andreypopp** [comment]
Just for the record: having the same issues with pnpm installer.

**Comment by ryyppy** [comment]
I tried some things... So when I  `npm install` locally in the checked out project, it works like a charm (node 4.4.0), because it uses `nodegit@0.13.2` ... if I install it via `npm install flow-typed -g`, it apparently fetches the wrong version of `nodegit` and fails... not sure how to fix this

**Comment by jeffmo** [comment]
@ryyppy: What happens if you `npm uninstall -g flow-typed` and then re-`npm install -g flow-typed`?

**Comment by ryyppy** [comment]
I really tried absolutely everything... I even did `rm -rf ~/.npm && npm cache clear` to make sure I don't use any funky caching in the background... I need more time to investigate... maybe node broke something with their `node-gyp` processes? who knows...

**Comment by ryyppy** [comment]
Oh dude.... it's `npm` v2 which causes problems... don't ask me why.
I did `npm install -g npm` and now it builds... not sure how we can deal with `npm v2` though

\cc @valscion

**Comment by jeffmo** [comment]
ohh, I think npm v2 doesn't auto-flatten (whereas v3 does). I'm about to send a PR for https://github.com/flowtype/flow-typed/issues/76 -- so I'll try bumping the minimum-viable version to `0.13.2` in package.json in that and see if it helps

**Comment by ryyppy** [comment]
That's a good idea,... this would also be useful to publish a new version `beta.6` , so we can provide the `update-cache` command for  #142

**Comment by valscion** [comment]
Very strange. Globally installing `npm install -g flow-typed@2.0.0-beta.5` still fails (I tried without the exact version specified, too), but if I `npm link` directly from the `cli` directory here, the manual compilation eventually succeeds. See gist for precise output: https://gist.github.com/valscion/aeb457470d524c8c71710ff46830cfc3

Maybe it has something to do with dependencies that are needed for the build and are listed only as `devDependencies`? That would explain why `npm link` succeeds, even if it needs a manual compilation, compared to a global installation.

**Comment by ryyppy** [comment]
@valscion Did you use `npm` >= v3?

**Comment by valscion** [comment]
No, I used the same npm version as I can't yet use npm3

**Comment by valscion** [comment]
Well thats fun.

If I add the needed `nodegit` packages, `npm install --save nopt npmlog@^2.0.0 rimraf tar`, inside the `cli` directory, then running `npm link` will not trigger a rebuild of `nodegit` from sources.

This is likely related to https://github.com/nodegit/nodegit/issues/1052

Maybe `nodegit` dependencies are botched?

---

Digging deeper, I found out that these packages should come from `node-pre-gyp` already. But as `node-pre-gyp` had at some point [incorrectly pointed out](https://github.com/mapbox/node-pre-gyp/commit/90b6eb7ca25008dd9ea565d1d221fc5dc815c7a7#commitcomment-17384452) that you should add  `node-pre-gyp` as `bundledDependencies`, and `nodegit` does indeed that, it might result in `node-pre-gyp` missing the required dependencies and thus causing all sorts of native compilation issues.

I'll dig deeper, see if I can resolve this issue all the way on `nodegit` level by modifying the package.json.

**Comment by valscion** [comment]
Nah still getting issues :( I can `npm link` from the `cli` directory, though, even without adding any packages I told above (so scrap that) but that also ends up compiling nodegit from the sources.

I even uninstalled my node installed via `nvm` and reinstalled it, but still the same issue prevails.

Seems like nodegit is having more issues than we thought: https://github.com/nodegit/nodegit/issues/1052

```
➜  cli git:(master) npm link
npm WARN deprecated minimatch@2.0.10: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
npm WARN deprecated minimatch@0.2.14: Please update to minimatch 3.0.2 or higher to avoid a RegExp DoS issue
/
> flow-bin@0.25.0 postinstall /Users/vesa/code/muut/flow-typed/cli/node_modules/flow-bin
> node lib/install.js

  ✔ flow binary test passed successfully

> jest-cli@11.0.0 postinstall /Users/vesa/code/muut/flow-typed/cli/node_modules/jest-cli
> node postinstall.js

|
> fsevents@1.0.12 install /Users/vesa/code/muut/flow-typed/cli/node_modules/babel-cli/node_modules/chokidar/node_modules/fsevents
> node-pre-gyp install --fallback-to-build

[fsevents] Success: "/Users/vesa/code/muut/flow-typed/cli/node_modules/babel-cli/node_modules/chokidar/node_modules/fsevents/lib/binding/Release/node-v46-darwin-x64/fse.node" is installed via remote
-
> nodegit@0.13.2 install /Users/vesa/code/muut/flow-typed/cli/node_modules/nodegit
> node lifecycleScripts/install

[nodegit] Fetching binary from S3.
[nodegit] Failed to install prebuilt binary:
{ [Error: Command failed: /bin/sh -c "/Users/vesa/code/muut/flow-typed/cli/node_modules/nodegit/node_modules/.bin/node-pre-gyp" install --fallback-to-build=false
module.js:327
    throw err;
    ^

Error: Cannot find module 'nopt'
    at Function.Module._resolveFilename (module.js:325:15)
    at Function.Module._load (module.js:276:25)
    at Module.require (module.js:353:17)
    at require (internal/module.js:12:17)
    at Object.<anonymous> (/Users/vesa/code/muut/flow-typed/cli/node_modules/nodegit/node_modules/node-pre-gyp/lib/node-pre-gyp.js:14:12)
    at Module._compile (module.js:409:26)
    at Object.Module._extensions..js (module.js:416:10)
    at Module.load (module.js:343:32)
    at Function.Module._load (module.js:300:12)
    at Module.require (module.js:353:17)
]
  killed: false,
  code: 1,
  signal: null,
  cmd: '/bin/sh -c "/Users/vesa/code/muut/flow-typed/cli/node_modules/nodegit/node_modules/.bin/node-pre-gyp" install --fallback-to-build=false' }
[nodegit] Building manually. (You'll be here a while.)
[nodegit] Regenerating and configuring code

--- clip lots of native build output ---
```

**Comment by jeffmo** [comment]
This is unfortunate :/ Given that nodegit 0.13 was a big update (and was relatively recent), maybe it's worth looking into switching back to v0.12.2 for a while while they get the kinks worked out.

I'll look into this this week.

**Comment by valscion** [comment]
Thanks! It's a shame that a downgrade is needed. I was a bit optimistic that we could've found the root cause but as I've banged my head against the wall with this one for a few hours already without making progress, or knowing where to dig deeper, I'm quite lost.

Hopefully nodegit will get these issues sorted out. I'm rooting for https://github.com/nodegit/nodegit/pull/1055 to make debugging this issue far easier, even fixing it if all goes well.

**Comment by johnhaley81** [comment]
@valscion I have nodegit/nodegit#1055 passing now. What would be the best way for me to test this in `flow-typed`?

**Comment by johnhaley81** [comment]
Hmm.. so digging into this more, it is exactly the `bundledDependencies` is indeed the problem. I'm adding that fix to the PR and in my local tests it seems to indeed fix the issue you guys are having.

**Comment by valscion** [comment]
Awesome, thanks a lot @johnhaley81! 😍

**Comment by valscion** [comment]
I can confirm that I can now install `flow-typed` after #162 landed, hooray! However, I still seem to have a nasty issue where I'm not able to install any libdefs, but that's another issue.
